sys/shell/commands/sc_gnrc_icmpv6_echo.c

```diff
@@ -0,0 +1,425 @@
+ * published under GPLv3
+ */
+
+#ifdef MODULE_GNRC_ICMPV6
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "bitfield.h"
+#include "byteorder.h"
+#include "kernel_types.h"
+#ifdef MODULE_LUID
+#include "luid.h"
+#endif
+#include "msg.h"
+#include "net/gnrc.h"
+#include "net/gnrc/icmpv6.h"
+#ifdef MODULE_GNRC_IPV6_NIB
+#include "net/gnrc/ipv6/nib/nc.h"
+#endif
+#include "net/icmpv6.h"
+#include "net/ipv6.h"
+#include "timex.h"
+#include "utlist.h"
+#include "xtimer.h"
+
+#define _SEND_NEXT_PING         (0xEF48)
+#define _PING_FINISH            (0xEF49)
+
+#define CKTAB_SIZE              (64U * 8)   /* 64 byte * 8 bit/byte */
+
+#define DEFAULT_COUNT           (3U)
+#define DEFAULT_DATALEN         (sizeof(uint32_t))
+#define DEFAULT_ID              (0x53)
+#define DEFAULT_INTERVAL_USEC   (1U * US_PER_SEC)
+#define DEFAULT_TIMEOUT_USEC    (1U * US_PER_SEC)
+
+typedef struct {
+    gnrc_netreg_entry_t netreg;
+    xtimer_t sched_timer;
+    msg_t sched_msg;
+    ipv6_addr_t host;
+    char *hostname;
+    unsigned long num_sent, num_recv, num_rept;
+    unsigned long long tsum;
+    unsigned tmin, tmax;
+    unsigned count;
+    size_t datalen;
+    BITFIELD(cktab, CKTAB_SIZE);
+    uint32_t timeout;
+    uint32_t interval;
+    kernel_pid_t iface;
+    uint16_t id;
+    uint8_t hoplimit;
+    uint8_t pattern;
+} _ping_data_t;
+
+static void _usage(char *cmdname);
+static int _configure(int argc, char **argv, _ping_data_t *data);
+static void _pinger(_ping_data_t *data);
+static void _print_reply(_ping_data_t *data, gnrc_pktsnip_t *icmpv6,
+                         ipv6_addr_t *from, unsigned hoplimit);
+static void _handle_reply(_ping_data_t *data, gnrc_pktsnip_t *pkt);
+static int _finish(_ping_data_t *data);
+
+int _gnrc_icmpv6_ping(int argc, char **argv)
+{
+    _ping_data_t data = {
+        .netreg = GNRC_NETREG_ENTRY_INIT_PID(ICMPV6_ECHO_REP,
+                                                 sched_active_pid),
+        .count = DEFAULT_COUNT,
+        .tmin = UINT_MAX,
+        .datalen = DEFAULT_DATALEN,
+        .timeout = DEFAULT_TIMEOUT_USEC,
+        .interval = DEFAULT_INTERVAL_USEC,
+        .id = DEFAULT_ID,
+        .pattern = DEFAULT_ID,
+    };
+    int res;
+
+    if ((res = _configure(argc, argv, &data)) != 0) {
+        return res;
+    }
+    gnrc_netreg_register(GNRC_NETTYPE_ICMPV6, &data.netreg);
+    _pinger(&data);
+    do {
+        msg_t msg;
+
+        msg_receive(&msg);
+        switch (msg.type) {
+            case GNRC_NETAPI_MSG_TYPE_RCV: {
+                _handle_reply(&data, msg.content.ptr);
+                gnrc_pktbuf_release(msg.content.ptr);
+                break;
+            }
+            case _SEND_NEXT_PING:
+                _pinger(&data);
+                break;
+            case _PING_FINISH:
+                goto finish;
+            default:
+                /* requeue wrong packets */
+                msg_send(&msg, sched_active_pid);
```

Comment:
What other messages can appear here? Shouldn't they be dropped instead of requeued? I can't see another `msg_recv` anywhere here that handles other types of messages, so the queue would get full with garbage ..

Reply:
I can research that, but on first glance I would say: remember that this is the main thread. Other application specific messages might land in the queue.

Reply:
(I did it because the current version in master does it as well: https://github.com/RIOT-OS/RIOT/blob/1d693403b65231b6784a25f6bb5a1c2657af6834/sys/shell/commands/sc_icmpv6_echo.c#L302-L305)

Reply:
>remember that this is the main thread. Other application specific messages might land in the queue.

yes, sounds sensible. Please research that anyway (:

Reply:
Ok, I had in memory, that this line existed because someone asked for it, but apparently I introduced it myself in https://github.com/RIOT-OS/RIOT/pull/2555. Anyway, given that this is `main` and that we didn't face any problems with this until now, I think the line is here correctly.